TickSpec/ServiceProvider.fs

```diff
@@ -3,21 +3,94 @@
 open System
 open System.Collections.Generic
 open System.Diagnostics
+open System.Reflection
 
+/// Provides an instance provider for tests
+type IInstanceProvider =
+    inherit IDisposable
+
+    /// Resolves an instance of desired type
+    abstract member Resolve : Type -> obj
+
+    /// Registers an instance for a type (if there is already a registered instance, it will be replaced)
+    abstract member RegisterInstance : Type -> obj -> unit
+
+[<Sealed>]
 /// Creates instance service provider
-type ServiceProvider () =
-    /// Type instances constructed for invoked steps
+type ServiceProvider () as self =
+    /// Type instances for invoked steps
     let instances = Dictionary<_,_>()
+
+    /// Resolves an instance for a specified type (and remembering the stack of types being resolved)
+    let rec resolveInstance (t:Type) (typeStack: Type list) =
+        let alreadyRequested =
+            typeStack
+            |> List.tryFind (fun x -> x = t)
+
+        match alreadyRequested with
+        | Some _ -> raise (InvalidOperationException(sprintf "Circular dependency found when resolving type %O" t))
+        | None -> ()
+
+        match t with
+        | t when t = typeof<IInstanceProvider> -> self :> obj
+        | t ->
+            match instances.TryGetValue t with
+            | true, instance -> instance
+            | false, _ ->
+                let instance = createInstance t typeStack
+                instances.Add(t, instance)
+                instance
+
+    /// Creates an instance if there was none for a speicified type (and remembering the stack of types being resolved)
+    and createInstance (t:Type) (typeStack: Type list) =
+        let constructors =
+            t.GetConstructors()
+            |> List.ofArray
+
+        let (_, widestConstructors) =

```

Comment:
am thinking, oh noes, we is writin' another DI container (though I agree that some form of batteries being included is important (how does this implf fare under .NET Core, did we say it was OK ?))

Reply:
If we move under .NET core then I would replace it by the one used in .NET Core. However, if you look on the implementation now, it is not DI, it is just the previous provider with recursion (so a type may need a different type) and with the ability to prepare an instance from outside. It seems to me ok, to use this approach., to have such simple container here and provide the ability to replace it by something real.

Reply:
Do you mean the one in ASP.NET Core ? Is that in a separable package ?

Reply:
Yes.

Reply:
Cool - that might make a pretty decent LCD container to target given the degree to which one should ideally be employing containers in an acceptance testing scenario.